//##########################################################################
//# Copyright (C) 2010-2012, Zenoss Inc.  All Rights Reserved.
//##########################################################################

import "zenoss/protocols/protobufs/model.proto";
import "zenoss/protocols/protobufs/util.proto";

package org.zenoss.protobufs.zep;

message EventActor {
    optional org.zenoss.protobufs.model.ModelElementType element_type_id = 1;
    optional string element_uuid = 2;
    optional string element_identifier = 3;
    optional string element_title = 4;

    optional org.zenoss.protobufs.model.ModelElementType element_sub_type_id = 5;
    optional string element_sub_uuid = 6;
    optional string element_sub_identifier = 7;
    optional string element_sub_title = 8;

    extensions 1000 to max;
}

message EventDetail {
    enum EventDetailMergeBehavior {
        REPLACE = 1; // Replace old values with new values
        APPEND = 2; // Append new values to old values
        UNIQUE = 3; // Append new values to old values and remove duplicate values
    }
    required string name = 1;
    repeated string value = 2;
    // When de-duping events or updating details on an existing event, if an existing
    // detail is found in the event with the same name, this enum controls how the
    // detail values should be merged.
    optional EventDetailMergeBehavior merge_behavior = 3 [default = REPLACE];

    extensions 1000 to max;
}

message EventDetailSet {
    repeated EventDetail details = 1;
}

// Tag events with arbitrary UUIDs. Tags are used for event rainbows and filtering.
message EventTag {
    required string type = 1;
    repeated string uuid = 2;
    extensions 1000 to max;
}

// Notes added to an event (formerly event log).
message EventNote {
    optional string uuid = 1;
    optional string user_uuid = 2;
    optional string user_name = 3;
    optional uint64 created_time = 4;
    required string message = 5;
    extensions 1000 to max;
    // extensions 1000-1099 reserved for GOM zenpack
}

enum EventSeverity {
    SEVERITY_CLEAR = 0;
    SEVERITY_DEBUG = 1;
    SEVERITY_INFO = 2;
    SEVERITY_WARNING = 3;
    SEVERITY_ERROR = 4;
    SEVERITY_CRITICAL = 5;
}

enum SyslogPriority {
    SYSLOG_PRIORITY_EMERG = 0;
    SYSLOG_PRIORITY_ALERT = 1;
    SYSLOG_PRIORITY_CRIT = 2;
    SYSLOG_PRIORITY_ERR = 3;
    SYSLOG_PRIORITY_WARNING = 4;
    SYSLOG_PRIORITY_NOTICE = 5;
    SYSLOG_PRIORITY_INFO = 6;
    SYSLOG_PRIORITY_DEBUG = 7;
}

enum EventStatus {
    STATUS_NEW = 0;
    STATUS_ACKNOWLEDGED = 1;
    STATUS_SUPPRESSED = 2;
    STATUS_CLOSED = 3; // Closed by the user.
    STATUS_CLEARED = 4; // Closed by a matching clear event.
    STATUS_DROPPED = 5; // Dropped via a transform.
    STATUS_AGED = 6; // Closed via automatic aging.
}

// Represents an individual event occurrence
message Event {
    optional string uuid = 1;
    optional uint64 created_time = 2;
    optional string fingerprint = 3;
    optional string event_class = 4;
    optional string event_class_key = 5;
    optional string event_class_mapping_uuid = 6;
    optional EventActor actor = 7;
    optional string summary = 8;
    optional string message = 9;
    optional EventSeverity severity = 10 [default = SEVERITY_INFO];
    optional string event_key = 12;
    optional string event_group = 13;
    optional string agent = 14;
    optional SyslogPriority syslog_priority = 15;
    optional uint32 syslog_facility = 16;
    optional uint32 nt_event_code = 17;
    optional string monitor = 18;
    repeated EventDetail details = 19;
    repeated EventTag tags = 20;
    optional string summary_uuid = 21;
    optional EventStatus status = 22 [default = STATUS_NEW];
    optional bool apply_transforms = 23 [default = true];
    // The incoming event count (if event was de-duped at the collector).
    optional uint32 count = 24 [default = 1];
    // The first seen time (if event was de-duped at the collector).
    optional uint64 first_seen_time = 25;

    // Event Flapping detection parameters, these are configured
    // in the zodb on event classes and added as an attribute to the event

    // The time interval that flapping_threshold number of flaps must occur in to be considered flapping (default 1 hour)
    optional uint32 flapping_interval_seconds = 26 [default = 3600];

    // The number of flaps an event set must have to be considered in a flapping state
    optional uint32 flapping_threshold = 27 [default = 4];

    // if an event is below this severity level, then the next event is above, it is considered a flap.
    optional EventSeverity flapping_severity = 28 [default = SEVERITY_ERROR];

    extensions 1000 to max;
    // extensions 1000-1099 are reserved for the GOM zenpack
}

// Represents an event that is sent to ZEP for processing.
// This is a bridge between zeneventd and ZEP
message ZepRawEvent {
    required Event event = 1;
    repeated string clear_event_class = 2;
    extensions 1000 to max;
}

// Represents a summary of an event (based on fingerprint).
message EventSummary {
    optional string uuid = 1;
    // Typically contains only the most recent occurrence.
    repeated Event occurrence = 2;
    optional EventStatus status = 3 [default = STATUS_NEW];
    optional uint64 first_seen_time = 4;
    optional uint64 status_change_time = 5;
    optional uint64 last_seen_time = 6;
    optional uint32 count = 7 [default = 1];
    optional string current_user_uuid = 8;
    optional string current_user_name = 9;
    optional string cleared_by_event_uuid = 10;
    repeated EventNote notes = 11;
    repeated EventAuditLog audit_log = 12;
    optional uint64 update_time = 13;
    extensions 1000 to max;
}

// Both from and to are inclusive. If 'from' is not specified, then 'from' is
// considered as negative infinity. If 'to' is not specified, then 'to' is
// considered positive infinity.
message NumberRange {
    optional sint64 from = 1;
    optional sint64 to = 2;
}

enum FilterOperator {
    OR = 1;
    AND = 2;
}

message EventTagFilter {
    optional FilterOperator op = 1 [default = OR];
    repeated string tag_uuids = 2;
    extensions 1000 to max;
}

// Values will be searched as an exact match unless otherwise
// specified by their detail type.
// Strings: Exact or wildcard if * or ?
// Numeric: Exact or range if in the form of M:, :N, or M:N
// IP Address: Exact or wildcard if * or ?, ranges in form of ADDR1-ADDR2, ADDR1/MASK
message EventDetailFilter {
    // All details in this filter share the FilterOperator specified.
    required string key = 1;
    repeated string value = 2;
    optional FilterOperator op = 3 [default = OR];
    extensions 1000 to max;
}

message EventFilter {
    repeated EventSeverity severity = 1;
    repeated EventStatus status = 2;

    // Accepted event_class queries:
    //
    //   /Class/NameFoo  -> Matches event_class "/Class/NameFoo" exactly.
    //   /Class/NameFoo/ -> Matches event_class "/Class/NameFoo" and any subclasses ("/Class/NameFoo/Bar" but not
    //                      "/Class/NameFooBar").
    //   /Class/NameFoo* -> Matches any event_class starting with "/Class/NameFoo".
    //   NameFoo         -> Matches any event_class with "NameFoo" as a component of the event class, i.e.
    //                      "/Class/NameFoo" or "/NameFoo" but not "/Class/NameFooBar".
    //   NameFoo*        -> Matches any event_class with a component that starts with "NameFoo", i.e. "/Class/NameFoo"
    //                      or "/NameFooBar".
    //   NameFoo/Bar     -> Matches any event class with the components in order "NameFoo" and "Bar". This matches
    //                      "/Class/NameFoo/Bar" but not "/Class/Bar/NameFoo".
    //   N*/B*           -> Matches any event class with components with prefix 'N' and 'B' in order.
    //
    // NOTE: All searches beginning with a '/' search the entire event class, otherwise the tokenized event class is
    //       searched.
    repeated string event_class = 3;

    repeated org.zenoss.protobufs.util.TimestampRange first_seen = 4;
    repeated org.zenoss.protobufs.util.TimestampRange last_seen = 5;
    repeated org.zenoss.protobufs.util.TimestampRange status_change = 6;
    repeated org.zenoss.protobufs.util.TimestampRange update_time = 7;
    repeated NumberRange count_range = 8;

    // Searches on element_identifier and element_sub_identifier provide optimized substring searching. To search the
    // full identifier, enclose your search in double quotes. For example:
    //
    //  host    -> Matches 'localhost', 'machine.myhostname.com'.
    //  "host"  -> Matches 'host' exactly.
    //  "host*" -> Matches all identifiers which start with 'host'.
    repeated string element_identifier = 9;
    repeated string element_sub_identifier = 10;

    repeated string uuid = 11; // event summary uuid

    // All search terms support wildcard queries ('*' and '?'). Phrases are supported within a query by enclosing
    // the phrase to be searched within double quotes.
    repeated string event_summary = 12;
    repeated string current_user_name = 13;
    repeated EventTagFilter tag_filter = 14;
    repeated EventDetailFilter details = 15;
    repeated string fingerprint = 16;
    repeated string agent = 17;
    repeated string monitor = 18;

    // operator is used for all fields on this filter. Repeateds are grouped and ORed.
    optional FilterOperator operator = 19 [default = AND];

    repeated EventFilter subfilter = 20;

    // Searches on element_title and element_sub_title use the same behavior as searching on element_identifier and
    // element_sub_identifier above.
    repeated string element_title = 21;
    repeated string element_sub_title = 22;
    repeated string event_key = 23;
    repeated string event_class_key = 24;
    repeated string event_group = 25;
    repeated string message = 26;
    extensions 1000 to max;
}

message EventSort {
    enum Field {
        SEVERITY = 1;
        STATUS = 2;
        EVENT_CLASS = 3;
        FIRST_SEEN = 4;
        LAST_SEEN = 5;
        STATUS_CHANGE = 6;
        COUNT = 7;
        ELEMENT_IDENTIFIER = 8;
        ELEMENT_SUB_IDENTIFIER = 9;
        EVENT_SUMMARY = 10;
        UPDATE_TIME = 11;
        CURRENT_USER_NAME = 12;
        AGENT = 13;
        MONITOR = 14;
        UUID = 15;
        FINGERPRINT = 16;
        DETAIL = 17; // detail key must be set, specified below.
        ELEMENT_TITLE = 18;
        ELEMENT_SUB_TITLE = 19;
        EVENT_KEY = 20;
        EVENT_CLASS_KEY = 21;
        EVENT_GROUP = 22;
    }

    enum Direction {
        ASCENDING = 1;
        DESCENDING = 2;
    }

    required Field field = 1;
    optional Direction direction = 2 [default = ASCENDING];
    optional string detail_key = 3;

    extensions 1000 to max;
}

// message for requesting event summaries.
message EventSummaryRequest {
    optional EventFilter event_filter = 1;
    optional EventFilter exclusion_filter = 2;
    repeated EventSort sort = 3;
    optional uint32 limit = 4 [default = 1000];
    optional uint32 offset = 5;

    extensions 1000 to max;
}

// message for response for EventSummaryRequest
message EventSummaryResult {
    repeated EventSummary events = 1;
    optional uint32 limit = 2;
    // Will be set to offset + limit if results remain, otherwise unset.
    optional uint32 next_offset = 3;
    optional uint32 total = 4;
    extensions 1000 to max;
}

// Properties which can be modified on an event summary
// This is passed with every EventSummaryUpdateRequest.
message EventSummaryUpdate {
    optional EventStatus status = 1;
    optional string current_user_uuid = 2;
    optional string current_user_name = 3;

    extensions 1000 to max;
    // extensions 1000-1099 are reserved for the GOM zenpack
}


// POST to events/search that starts a saved search.
// Returns a 204 created if successful.
// DELETE to the uri in order to expire the search before the timeout.
message EventQuery {
    optional EventFilter event_filter = 1;
    optional EventFilter exclusion_filter = 2;
    repeated EventSort sort = 3;
    optional uint32 timeout = 4 [default=60];

    extensions 1000 to max;
}

// once a query has been established, you will have the event_query_uuid
// with which you will need to make subsequent update requests.
message EventSummaryUpdateRequest {
    optional string event_query_uuid = 1;
    required EventSummaryUpdate update_fields = 2;
    optional uint32 offset = 3 [default = 0];
    optional uint32 limit = 4 [default = 100];

    extensions 1000 to max;
}

message EventSummaryUpdateResponse {
    // Message used to update the next sequence in the collection.
    // This is unset if there are no events left to update.
    optional EventSummaryUpdateRequest next_request = 1;
    optional uint32 total = 2; // Never changes - just informational.
    required uint32 updated = 3; // The number of events updated in this batch.

    extensions 1000 to max;
}

message EventDetailItem {
    enum EventDetailType {
        STRING = 1;
        INTEGER = 2;
        FLOAT = 3;
        LONG = 4;
        DOUBLE = 5;
        IP_ADDRESS = 6;
        // A slash-delimited path to a resource (event classes, locations, systems, groups, device classes).
        PATH = 7;
    }
    required string key = 1;
    optional EventDetailType type = 2 [default = STRING];
    optional string name = 3;

    extensions 1000 to max;
}

// GET on some zep uri will return this proto as a response
message EventDetailItemSet {
    repeated EventDetailItem details = 1;
}

// Event Triggers
enum RuleType {
    RULE_TYPE_JYTHON = 1;
}

// The content is evaluated by a "ScriptEngine" for RuleType and checked against the event
message Rule {
    required int32 api_version = 1;
    required RuleType type = 2;
    required string source = 3;

    extensions 1000 to max;
}

message EventTriggerSubscription {
    optional string uuid = 1;
    // Wait this amount of time for a possible clear before sending a signal
    optional int32 delay_seconds = 2;
    // After first signal, wait this amount of time before sending a signal again (0 for no repeating)
    optional int32 repeat_seconds = 3;
    // should we send a signal on every occurrence of an event or only on the first
    optional bool send_initial_occurrence = 4 [default = true];
    // id of subscriber
    required string subscriber_uuid = 5;
    // id of trigger to subscribe to
    required string trigger_uuid = 6;

    extensions 1000 to max;
}

message EventTriggerSubscriptionSet {
    repeated EventTriggerSubscription subscriptions = 1;
}

message EventTrigger {
    optional string uuid = 1;
    optional string name = 2;
    optional bool enabled = 3 [default = true];

    // If rule matches event, start responding to subscriptions
    required Rule rule = 4;

    repeated EventTriggerSubscription subscriptions = 5;

    extensions 1000 to max;
}

message EventTriggerSet {
    repeated EventTrigger triggers = 1;
}

// A signal is sent because a Trigger rule matched an event and we
// waited for a EventTriggerSubscription.delay_seconds amount of time before
// sending the signal. Action is taken immediately for any signal received.
message Signal {
    required string uuid = 1;
    required uint64 created_time = 2;
    required string trigger_uuid = 3;
    required string subscriber_uuid = 4;
    // If this is signal clears all previous signals
    optional bool clear = 5 [default = false];
    optional EventSummary event = 6;
    optional EventSummary clear_event = 7;
    optional string message = 8;

    extensions 1000 to max;
}

message EventTagSeverity {
    required EventSeverity severity = 1;
    optional uint32 count = 2 [default=0];
    optional uint32 acknowledged_count = 3 [default=0];

    extensions 1000 to max;
}

message EventTagSeverities {
    required string tag_uuid = 1;
    repeated EventTagSeverity severities = 2;
    // The total number of event occurrences for this tag_uuid
    optional uint32 total = 3;
}

message EventTagSeveritiesSet {
    repeated EventTagSeverities severities = 1;
}

message EventAuditLog {
    required uint64 timestamp = 1;
    required EventStatus new_status = 2;
    optional string user_uuid = 3;
    optional string user_name = 4;

    extensions 1000 to max;
}

message ZepConfig {
    // Events with a lower severity than this are selected for aging (along
    // with event_age_interval_minutes). If you specify SEVERITY_CLEAR for
    // this setting, it disables event aging.
    optional EventSeverity event_age_disable_severity = 1 [default = SEVERITY_ERROR];

    // The number of minutes that elapse from the last time an event was seen
    // until it is a candidate for event aging. Setting this to 0 disables
    // event aging.
    optional uint32 event_age_interval_minutes = 2 [default = 240];

    // The number of minutes since a closed event was last seen before it is moved
    // to the event archive. The minimum value is 1 and the maximum value is 43200.
    optional uint32 event_archive_interval_minutes = 3 [default = 4320];

    // The number of days that events in the event archive are saved. By default
    // they are kept in the archive for 90 days. The minimum value is 1 and the
    // maximum value is determined by the range of event archive partitions. With
    // the default configuration, the maximum value is 1000 days.
    optional uint32 event_archive_purge_interval_days = 4 [default = 90];

    // The number of days that event occurrence times are kept. By default
    // they are kept for 1 day. The minimum value is 1 and the
    // maximum value is determined by the range of event time partitions. With
    // the default configuration, the maximum value is 7 days.
    optional uint32 event_time_purge_interval_days = 5 [default = 1];

    // If we want to include the selected severity as well as the
    // severties below it when aging
    optional bool event_age_severity_inclusive = 6 [default = false];

    // The maximum size of an event that will be processed in bytes.
    // Events that are too large will be logged and dropped.
    // Events that will *become* too big will have their details overwritten
    // with new details. The minimum value is 8192, the maximum value
    // is 102400.
    optional uint64 event_max_size_bytes = 7 [default = 32768];

    // The default indexing interval of the event summary (in milliseconds).
    optional uint64 index_summary_interval_milliseconds = 8 [default = 1000];

    // The default indexing interval of the event archive (in milliseconds).
    optional uint64 index_archive_interval_milliseconds = 9 [default = 30000];

    // The number of events to index each index interval_milliseconds
    // interval.
    optional uint32 index_limit = 10 [default = 1000];

    // The maximum number of events to age in each interval. This limit should
    // be kept relatively low to prevent large database transactions.
    optional uint32 aging_limit = 11 [default = 1000];

    // The maximum number of events to archive in each interval. This limit
    // should be kept relatively low to prevent large database transactions.
    optional uint32 archive_limit = 12 [default = 1000];

    // The interval when events are scanned to perform auto-aging.
    optional uint64 aging_interval_milliseconds = 13 [default = 60000];

    // The interval when events are scanned for moving to the archive.
    optional uint64 archive_interval_milliseconds = 14 [default = 60000];

    // If we want to enable event flapping detection
    optional bool enable_event_flapping_detection = 15 [default = false];

    // The event class that we want the generated flapping event to be
    optional string flapping_event_class = 19 [default = "/Status/Flapping"];
    extensions 1000 to max;
}

message DaemonHeartbeat {
    // The collector sending the heartbeat.
    required string monitor = 1;

    // The name of the daemon running on the collector.
    required string daemon = 2;

    // The number of seconds this heartbeat is valid for before heartbeat
    // failures are sent for the daemon.
    required uint32 timeout_seconds = 3;

    // The time the heartbeat was last received.
    optional uint64 last_time = 4;

    extensions 1000 to max;
}

message DaemonHeartbeatSet {
    repeated DaemonHeartbeat heartbeats = 1;
}

message EventTime {
    optional string summary_uuid = 1;
    optional uint64 processed_time = 2;
    optional uint64 created_time = 3;
    optional uint64 first_seen_time = 4;
}

message EventTimeSet {
    repeated EventTime event_times = 1;
}

message ZepStatistic {
    required string name = 1;
    required string description = 2;
    required int64 value = 3;
    extensions 1000 to max;
}

message ZepStatistics {
    repeated ZepStatistic stats = 1;
    extensions 1000 to max;
}
